hooks/useProtocolImport.tsx

```diff
@@ -160,9 +162,47 @@ export const useProtocolImport = () => {
       }
 
       const assets = await getProtocolAssets(protocolJson, zip);
-      let assetsWithCombinedMetadata: z.infer<typeof assetInsertSchema> = [];
+
+      const newAssets: typeof assets = [];
+
+      const assetsWithCombinedMetadata: z.infer<typeof assetInsertSchema> = [];
+
+      let newAssetsWithCombinedMetadata: z.infer<typeof assetInsertSchema> = [];
+
+      // Check if the assets are already in the database.
+      // If yes, add them to assetsWithCombinedMetadata to be connected to the protocol.
+      // If not, add them to newAssets to be uploaded.
 
       if (assets.length > 0) {
+        const assetPromises = assets.map(async (asset) => {
```

Comment:
I _kind of_ get why you did it this way, but I also think we should make this be more efficient. 

Right now you're creating a separate query for every asset in the protocol. You are then immediately invoking all of these queries, and waiting for them all to settle. There are a few problems with that:

1. All the queries are competing with each other, since they are all immediately invoked. This means all network requests will be sent at the same time, regardless of what else the client or the server are doing. This will lead to performance issues eventually, and is almost always not what you intended. If you wanted to do it this way, a queue with some managed concurrency would be better.
2. The way you've structured this with a `promise.all` means that any individual failure with any of these (many, many!) queries will abort the whole import. Its basically impossible to do any error handling.
3. One invokation of the tRPC endpoint per-asset is going to be very resource intensive on the database server, too. It will have to read all rows in the table for each query, and subsequent queries won't be able to benefit from any information from previous ones.

Here's what I suggest:
1. Refactor the router to accept an array of IDs to check, and return something useful like the IDs that don't already exist. If you want it to be a more general-purpose endpoint, you could return either an object with keys corresponding to each array item and a boolean value, OR something like `{ found: [123,456], notFound: [789] }`.
2. Call this endpoint inside a try/catch block with all protocol assets in one go, so that you can handle the error case that this check failed.
3. Tidy up the code following this - I don't like the look of L:195!

Reply:
refactored using this approach here [a3be68e](https://github.com/complexdatacollective/Fresco/pull/93/commits/a3be68e49cdca722f050a9a157639cfdd0f4cf87)